subprojects/core/src/main/java/org/gradle/cache/internal/CacheRepositoryServices.java

```diff
@@ -35,7 +35,7 @@ public CacheRepositoryServices(File gradleUserHomeDir, @Nullable File projectCac
         this.projectCacheDir = projectCacheDir;
     }
 
-    protected CacheScopeMapping createCacheScopeMapping() {
+    protected DefaultCacheScopeMapping createCacheScopeMapping() {
```

Comment:
`CacheScopeMapping` does not implement `GlobalCache`, which `DefaultCacheScopeMapping` does. Our injection framework only seems to detect interfaces implemented by the type returned from the creation method, so we need to return `DefaultCacheScopeMapping` here so that it is picked up when injecting `List<GlobalCache>`.

Reply:
That is weird. Shouldn't we fix that problem?

Reply:
Ah, I think I understand why the interfaces need to be part of the return type: They need to be known at compilation time, since the actual creation method will not be invoked when it doesn't provide any required service. So I don't see a way how to fix this.

Reply:
I think I would keep the `GlobalCache` thing from the interface, since it seems to be something internal which shouldn't be used by clients of `CacheScopeMapping`.

Reply:
I still don't like how error-prone this is. We should at least add a comment here saying why we are deviating from the typical way of returning the interface.

Reply:
I added a comment.